### PR TITLE
add random foreground crop for boxes

### LINKED ISSUE #3572: Box transforms
**Is your feature request related to a problem? Please describe.**
Box pre- and post-processing transform functions are not in MONAI

**Describe the solution you'd like**
Add box transform functions

**Describe alternatives you've considered**
N/A

**Additional context**
1. convert input box with original mode to the box with standard mode
2. convert output box with standard mode to the box with original mode
3. Flip boxes with image as a transform

### PR DESCRIPTION
**Pull request by Can-Zhao**
Signed-off-by: Can Zhao <canz@nvidia.com>

Fixes #3572  .

### Description
add random foreground crop for boxes

### Status
**Ready/Work in progress/Hold**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [ ] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [x] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

File: monai/apps/detection/transforms/array.py [hunk 1]
```diff
@@ -452,3 +444,102 @@ def __call__(self, boxes_mask: NdarrayOrTensor) -> Tuple[NdarrayOrTensor, Ndarra
             - classification foreground(fg) labels, dtype should be int, sized (N,).
         """
         return convert_mask_to_box(boxes_mask, self.bg_label, self.box_dtype, self.label_dtype)
+
+
+class SpatialCropPadBox(Transform):
```
Comment (Nic-Ma) [inline monai/apps/detection/transforms/array.py:449]: Is it possible to inherit from the `SpatialCrop` transform?
I see much duplicated code.

Thanks.

File: monai/apps/detection/transforms/box_ops.py [hunk 2]
```diff
@@ -318,3 +318,31 @@ def convert_mask_to_box(
     boxes, *_ = convert_to_dst_type(src=boxes_np, dst=boxes_mask, dtype=box_dtype)
     labels, *_ = convert_to_dst_type(src=labels_np, dst=boxes_mask, dtype=label_dtype)
     return boxes, labels
+
+
+def select_labels(
+    labels: Union[Sequence[NdarrayOrTensor], NdarrayOrTensor], keep: NdarrayOrTensor
+) -> Union[Tuple, NdarrayOrTensor]:
+    """
+    For element i in labels, select labels[i][keep]
```
Comment (Nic-Ma) [inline monai/apps/detection/transforms/box_ops.py:327]: Could you please help add more description for the function?
I feel this doc-string is almost the same as one line source code..`For element i in labels, select labels[i][keep]`

Thanks.

File: monai/apps/detection/transforms/dictionary.py [hunk 3]
```diff
@@ -935,6 +941,252 @@ def __call__(self, data: Mapping[Hashable, NdarrayOrTensor]) -> Dict[Hashable, N
...
+            the requested ROI in any dimension. If `True`, any smaller dimensions will be set to
+            match the cropped size (i.e., no cropping in that dimension).
+        mode: available padding modes for numpy array:{``"constant"``, ``"edge"``, ``"linear_ramp"``, ``"maximum"``,
+                ``"mean"``, ``"median"``, ``"minimum"``, ``"reflect"``, ``"symmetric"``, ``"wrap"``, ``"empty"``}
+                available modes for PyTorch Tensor: {``"constant"``, ``"reflect"``, ``"replicate"``, ``"circular"``}.
+                One of the listed string values or a user supplied function. Defaults to ``"constant"``.
+                See also: https://numpy.org/doc/1.18/reference/generated/numpy.pad.html
+                https://pytorch.org/docs/stable/generated/torch.nn.functional.pad.html
+                it also can be a sequence of string, each element corresponds to a key in ``keys``.
+        allow_missing_keys: don't raise exception if key is missing.
```
Comment (Nic-Ma) [inline monai/apps/detection/transforms/dictionary.py:989]: seems missing the `pad_kwargs` for pad function.

Thanks.

**Review by Nic-Ma — APPROVED** [review]
Thanks for the quick update.
It overall looks good to me now.
Put some minor comments inline.

Thanks.

File: monai/apps/detection/transforms/box_ops.py [hunk 4]
```diff
@@ -318,3 +318,36 @@ def convert_mask_to_box(
...
+        labels: Sequence of array. Each element represents classification labels or scores
+            corresponding to ``boxes``, sized (N,).
+        keep: the indices to keep, same length with each element in labels.
+
+    Return:
+        selected labels, does not share memory with original labels.
+    """
+    if isinstance(labels, torch.Tensor) or isinstance(labels, np.ndarray):
+        # will fail if simply use ensure_tuple(labels) for all cases.
+        labels_tuple = (labels,)
```
[the remark this answers is not in the document]
Comment (Can-Zhao) [reply monai/apps/detection/transforms/box_ops.py:339]: `ensure_tuple(torch.tensor([]))` will return `()`, instead of `(torch.tensor([]),)`

File: monai/apps/detection/transforms/box_ops.py [hunk 5]
```diff
@@ -318,3 +318,36 @@ def convert_mask_to_box(
...
+    else:
+        labels_tuple = ensure_tuple(labels)  # type: ignore
+
+    labels_select_list = []
+    keep_t: torch.Tensor = convert_data_type(keep, torch.Tensor)[0]
+    for i in range(len(labels_tuple)):
+        labels_t: torch.Tensor = convert_data_type(labels_tuple[i], torch.Tensor)[0]
+        labels_t = deepcopy(labels_t[keep_t, ...])
```
[the remark this answers is not in the document]
Comment (Can-Zhao) [reply monai/apps/detection/transforms/box_ops.py]: I changed it to `labels_t = labels_t[keep_t, ...].clone()` Not sure whether should detach...

[the remark this answers is not in the document]
Comment (Can-Zhao) [reply monai/apps/detection/transforms/box_ops.py]: tuple should be immutable...I'm silly
